source/sav/Sav3.cpp

```diff
@@ -0,0 +1,792 @@
+    std::array<int, BLOCK_COUNT> o2     = getBlockOrder(dt, 0xE000);
+    int activeSAV                       = getActiveSaveIndex(dt, o1, o2);
+    std::array<int, BLOCK_COUNT>& order = activeSAV == 0 ? o1 : o2;
+
+    int ABO = activeSAV * SIZE_BLOCK * 0xE;
+
+    int blockOfs0 = ((std::find(order.begin(), order.end(), 0) - order.begin()) * SIZE_BLOCK) + ABO;
+
+    // Get version
+    u32 gameCode = Endian::convertTo<u32>(&dt[blockOfs0 + 0xAC]);
+    switch (gameCode)
+    {
+        case 1:
+            return Game::FRLG; // fixed value
+        case 0:
+            return Game::RS; // no battle tower record data
+        default:
+            // Ruby doesn't set data as far down as Emerald.
+            // 00 FF 00 00 00 00 00 00 00 FF 00 00 00 00 00 00
+            // ^ byte pattern in Emerald saves, is all zero in Ruby/Sapphire as far as I can tell.
+            // Some saves have had data @ 0x550
+            if (Endian::convertTo<u64>(&dt[blockOfs0 + 0xEE0]) != 0)
+                return Game::E;
+            if (Endian::convertTo<u64>(&dt[blockOfs0 + 0xEE8]) != 0)
+                return Game::E;
+            return Game::RS;
+    }
+}
+
+Sav3::Sav3(std::shared_ptr<u8[]> dt) : Sav(dt, 0x20000)
+{
+    loadBlocks();
+
+    // Japanese games are limited to 5 character OT names; any unused characters are 0xFF.
+    // 5 for JP, 7 for INT. There's always 1 terminator, thus we can check 0x6-0x7 being 0xFFFF = INT
+    // OT name is stored at the top of the first block.
+    japanese = Endian::convertTo<s16>(&data[blockOfs[0] + 0x6]) == 0;
+
+    PokeDex = blockOfs[0] + 0x18;
+
+    initialize();
+}
+
+void Sav3::initialize(void)
+{
+    // Set up PC data buffer
+    Box = std::shared_ptr<u8[]>(new u8[SIZE_RESERVED]);
+
+    // Copy chunk to the allocated location
+    for (int i = 5; i < BLOCK_COUNT; i++)
+    {
+        unsigned int blockIndex = std::find(blockOrder.begin(), blockOrder.end(), i) - blockOrder.begin();
+        if (blockIndex == blockOrder.size()) // block empty
+            continue;
+        memcpy(Box.get() + ((i - 5) * 0xF80), data.get() + (blockIndex * SIZE_BLOCK) + ABO(), chunkLength[i]);
+    }
+
+    // LoadEReaderBerryData();
+
+    // Sanity Check SeenFlagOffsets -- early saves may not have block 4 initialized yet
+    std::vector<int> seenFlagOffsetsTemp;
+    for (auto seenFlagOffset : seenFlagOffsets)
+    {
+        if (seenFlagOffset >= 0)
+            seenFlagOffsetsTemp.push_back(seenFlagOffset);
+    }
+    seenFlagOffsets = seenFlagOffsetsTemp;
+}
+
+void Sav3::resign(void)
+{
+    // Copy Box data back
+    for (int i = 5; i < BLOCK_COUNT; i++)
+    {
+        unsigned int blockIndex = std::find(blockOrder.begin(), blockOrder.end(), i) - blockOrder.begin();
+        if (blockIndex == blockOrder.size()) // block empty
+            continue;
+        memcpy(data.get() + (blockIndex * SIZE_BLOCK) + ABO(), Box.get() + ((i - 5) * 0xF80), chunkLength[i]);
+    }
```

Comment:
Style thing: I'd rather this be in `Sav3::encrypt()` rather than `Sav3::resign()`, as it's closer to an actual encryption process than it is to a resigning operation. Might change `Sav::encrypt()` to `Sav::finalize()` to reflect that after merging

Reply:
Should I just remove `Sav3::resign()` completely and rename this to `Sav3::encrypt()`? It seems to build fine without `Sav3::resign()`

Reply:
`Sav3::setChecksums` should be renamed to `Sav3::resign` (because that's what it's doing)

Reply:
Ah, yeah